frontend/packages/pipelines-plugin/integration-tests/support/step-definitions/pipelines/pipelines-runs.ts

```diff
@@ -473,3 +479,108 @@ Then(
     modal.shouldBeClosed();
   },
 );
+
+Given('user is at pipeline page in developer perspective', () => {
+  perspective.switchTo(switchPerspective.Developer);
+});
+
+Given('a failed pipeline is present', () => {
+  navigateTo(devNavigationMenu.Add);
+  addPage.selectCardFromOptions(addOptions.ImportFromGit);
+  gitPage.enterGitUrl('https://github.com/sclorg/golang-ex');
+  devFilePage.verifyValidatedMessage('https://github.com/sclorg/golang-ex');
+  gitPage.selectAddPipeline();
+  gitPage.clickCreate();
+  topologyPage.verifyTopologyPage();
+});
+
+When('user goes to failed pipeline run of pipeline {string}', (pipelineName: string) => {
+  navigateTo(devNavigationMenu.Pipelines);
+  pipelinesPage.search(pipelineName);
+  cy.get(`[data-test-id^="${pipelineName}-"]`).click();
+});
+
+When('user opens pipeline run details', () => {
+  cy.get('.pf-c-breadcrumb').should('include.text', 'PipelineRun details');
+});
+
+Then('user can see status as Failure', () => {
+  cy.get(pipelineRunsPO.pipelineRunsTable.status).should('include.text', 'Failed');
+});
+
+Then('user can view failure message under Message heading', () => {
+  cy.get('.odc-pipeline-run-details__customDetails').within(() => {
```

Comment:
this selector would look better in a page object, actually lets get as many selectors that aren't parameterized into a page object as we can